xarray/coding/variables.py

```diff
@@ -189,8 +189,36 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype):
     return data
 
 
-def _choose_float_dtype(dtype, has_offset):
+def _choose_float_dtype(dtype, scale_factor, add_offset):
     """Return a float dtype that can losslessly represent `dtype` values."""
+    # Implementing cf-convention
+    # http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch08.html:
+    # Detail:
+    # If the scale_factor and add_offset attributes are of the same
+    # data type as the
+    # associated variable, the unpacked data is assumed to be of the same
+    # data type as the packed data. However, if the scale_factor
+    # and add_offset attributes are of a different data type
+    # from the variable (containing the packed data)
+    # then the unpacked data should match
+    # the type of these attributes, which must both be of type float or both
+    # be of type double. An additional restriction in this case is that
```

Comment:
What should we do if `add_offset`/`scale_factor` are *not* floats?

Right now in your implementation, if `scale_factor` is mistakenly `np.int64(10)` then the data would get decoded as int64, which seems like a bad idea (e.g., if the data is some float type). I think we would should probably ignore the type of `scale_factor`/`add_offset` in these cases instead.

Reply:
Hi @shoyer! Sorry for my delayed intervenience... First of all I'd like to thank and congratulate you for your efficiency and say that I'm very glad to see that you've taken into account my considerations about this topic.
About the comment above I'd say that, since the _CFScaleOffsetCoder_ will always multiply the raw data by the scale_factor, in case the scale_factor is an integer the decoded values will always result into integers too. Unless the raw data dtype is float, right?
Even in these cases I'd return the decoded values with the same dtype as the scale_factor, which is the expected behavior. (data providers must be aware... and you can always call the _decode_cf_variable_ function later (after using the _values.astype_('whatever') or changing the encoding['scale_factor'] dtype...).

By the way, I've also noted that the _CFScaleOffsetCoder.encode_ changes the _values.dtype_ before applying the inverse operation (i.e the dividing by the scale_factor), this will result in truncated values. I've already faced this situation, with my own netcdf encoding processes, and would advice you to only change the dtype after applying the transformation and use the numpy.round function (for integers encoding) to avoid the truncation artifacts.

Once again, sorry for my extensive comments and not opening an issue, as I should, but my hope is that you evaluate my considerations and take your own decisions with your colleagues, and keep the nice work you all have being done.  👍 

Cheers!